mypy: add type checking

aciba90 opened the pull request:
Enable mypy type checking and add more type definitions.

Fix bug in OciInstance.add_network_interface, discovered
by mypy. In a previous refactoring, we factor out the
oci.utils.get_subnet_id function to be used in
oci.cloud and oci.instance. This function does need the
availability_domain but OciInstance had not that attribute.
availability_domain added to OciInstance and fix the bug.

blackboxsw reviewed (approved):
Thanks Alberto good to get this cleaned up and operational in CI.

1. Minor typing additions requested inline for any signature we touched. (diff will follow)
2. Please also add a bit more context in your squashed commit message about what is being done and why in Oci.add_network_interface.

"fix OciInstance.add_network_interface" doesn't give enough content for the future us.

Approved when either way.

blackboxsw commented:
```diff
diff --git a/pycloudlib/azure/cloud.py b/pycloudlib/azure/cloud.py
index e6f6203..644ebe7 100644
--- a/pycloudlib/azure/cloud.py
+++ b/pycloudlib/azure/cloud.py
@@ -58,14 +58,14 @@ class Azure(BaseCloud):
     def __init__(
         self,
         tag,
-        timestamp_suffix=True,
+        timestamp_suffix: Optional[bool] = True,
         config_file: Optional[ConfigFile] = None,
         *,
-        client_id=None,
-        client_secret=None,
-        subscription_id=None,
-        tenant_id=None,
-        region=None,
+        client_id: Optional[str] = None,
+        client_secret: Optional[str] = None,
+        subscription_id: Optional[str] = None,
+        tenant_id: Optional[str] = None,
+        region: Optional[str] = None,
     ):
         """Initialize the connection to Azure.
 
@@ -535,7 +535,7 @@ class Azure(BaseCloud):
         raise ValueError("Invalid image_type")
 
     def daily_image(
-        self, release, *, image_type: ImageType = ImageType.GENERIC, **kwargs
+        self, release: str, *, image_type: ImageType = ImageType.GENERIC, **kwargs
     ):
         """Find the image info for the latest daily image for a given release.
 
diff --git a/pycloudlib/cloud.py b/pycloudlib/cloud.py
index 59e3e16..2d630ee 100644
--- a/pycloudlib/cloud.py
+++ b/pycloudlib/cloud.py
@@ -34,8 +34,8 @@ class BaseCloud(ABC):
 
     def __init__(
         self,
-        tag,
-        timestamp_suffix=True,
+        tag: str,
+        timestamp_suffix: Optional[bool] = True,
         config_file: Optional[ConfigFile] = None,
         required_values=None,
     ):
diff --git a/pycloudlib/ec2/cloud.py b/pycloudlib/ec2/cloud.py
index 4d47ca2..c07e04a 100644
--- a/pycloudlib/ec2/cloud.py
+++ b/pycloudlib/ec2/cloud.py
@@ -20,13 +20,13 @@ class EC2(BaseCloud):
 
     def __init__(
         self,
-        tag,
-        timestamp_suffix=True,
+        tag: str,
+        timestamp_suffix: Optional[bool] = True,
         config_file: Optional[ConfigFile] = None,
         *,
-        access_key_id=None,
-        secret_access_key=None,
-        region=None,
+        access_key_id: Optional[str] = None,
+        secret_access_key: Optional[str] = None,
+        region: Optional[str] = None,
     ):
         """Initialize the connection to EC2.
 
@@ -92,9 +92,9 @@ class EC2(BaseCloud):
 
     def released_image(
         self,
-        release,
+        release: str,
         *,
-        arch="x86_64",
+        arch: str = "x86_64",
         image_type: ImageType = ImageType.GENERIC,
         **kwargs,
     ):
@@ -190,9 +190,9 @@ class EC2(BaseCloud):
 
     def daily_image(
         self,
-        release,
+        release: str,
         *,
-        arch="x86_64",
+        arch: str = "x86_64",
         image_type: ImageType = ImageType.GENERIC,
         **kwargs,
     ):
diff --git a/pycloudlib/gce/cloud.py b/pycloudlib/gce/cloud.py
index 5b325f8..3792dc2 100644
--- a/pycloudlib/gce/cloud.py
+++ b/pycloudlib/gce/cloud.py
@@ -30,15 +30,15 @@ class GCE(BaseCloud):
 
     def __init__(
         self,
-        tag,
-        timestamp_suffix=True,
+        tag: str,
+        timestamp_suffix: Optional[bool] = True,
         config_file: Optional[ConfigFile] = None,
         *,
-        credentials_path=None,
-        project=None,
-        region=None,
-        zone=None,
-        service_account_email=None,
+        credentials_path: Optional[str] = None,
+        project: Optional[str] = None,
+        region: Optional[str] = None,
+        zone: Optional[str] = None,
+        service_account_email: Optional[str] = None,
     ):
         """Initialize the connection to GCE.
 
@@ -230,9 +230,9 @@ class GCE(BaseCloud):
 
     def daily_image(
         self,
-        release,
+        release: str,
         *,
-        arch="x86_64",
+        arch: str = "x86_64",
         image_type: ImageType = ImageType.GENERIC,
         **kwargs,
     ):
diff --git a/pycloudlib/openstack/cloud.py b/pycloudlib/openstack/cloud.py
index 90ef576..a9afc7e 100644
--- a/pycloudlib/openstack/cloud.py
+++ b/pycloudlib/openstack/cloud.py
@@ -17,11 +17,11 @@ class Openstack(BaseCloud):
 
     def __init__(
         self,
-        tag,
-        timestamp_suffix=True,
+        tag: str,
+        timestamp_suffix: Optional[bool] = True,
         config_file: Optional[ConfigFile] = None,
         *,
-        network=None,
+        network: Optional[str] = None,
     ):
         """Initialize the connection to openstack.

```

aciba90 commented:
> Thanks Alberto good to get this cleaned up and operational in CI.
> 
>     1. Minor typing additions requested inline for any signature we touched. (diff will follow)
> 
>     2. Please also add a bit more context in your squashed commit message about what is being done and why in Oci.add_network_interface.
> 
> 
> "fix OciInstance.add_network_interface" doesn't give enough content for the future us.
> 
> Approved when either way.

Thanks Chad for the review and suggestions!

I have addressed your comments, added extra type definitions and improved the squashed commit message.
